arch/x86/crc32_fold_pclmulqdq.c

```diff
@@ -380,6 +386,133 @@ Z_INTERNAL void crc32_fold_copy_pclmulqdq(crc32_fold *crc, uint8_t *dst, const u
     crc32_fold_save_partial((__m128i *)crc->fold, xmm_crc_part);
 }
 
+#define ONCE(op) if (first) { \
+    first = 0; \
+    (op); \
+}
+#define XOR_INITIAL(where) ONCE(where = _mm_xor_si128(where, xmm_initial))
+
+Z_INTERNAL void crc32_fold_pclmulqdq(crc32_fold *crc, const uint8_t *src, size_t len, uint32_t init_crc) {
+    unsigned long algn_diff;
+    __m128i xmm_t0, xmm_t1, xmm_t2, xmm_t3;
+    __m128i xmm_crc0, xmm_crc1, xmm_crc2, xmm_crc3, xmm_crc_part;
+    __m128i xmm_initial = _mm_cvtsi32_si128(init_crc);
+    char ALIGNED_(16) partial_buf[16] = { 0 };
+    int first = 1;
+
+    crc32_fold_load((__m128i *)crc->fold, &xmm_crc0, &xmm_crc1, &xmm_crc2, &xmm_crc3);
+
+    if (len < 16) {
+        if (len == 0)
+            return;
+
+        memcpy(partial_buf, src, len);
+        xmm_crc_part = _mm_load_si128((const __m128i *)partial_buf);
+        XOR_INITIAL(xmm_crc_part);
+        goto partial_nocopy;
+    }
+
+    algn_diff = ((uintptr_t)16 - ((uintptr_t)src & 0xF)) & 0xF;
+    if (algn_diff) {
+        xmm_crc_part = _mm_loadu_si128((__m128i *)src);
+
+        src += algn_diff;
+        len -= algn_diff;
+
+        XOR_INITIAL(xmm_crc_part);
+        partial_fold(algn_diff, &xmm_crc0, &xmm_crc1, &xmm_crc2, &xmm_crc3, &xmm_crc_part);
+    } else {
+        xmm_crc_part = _mm_setzero_si128();
```

Comment:
Kinda might work but we'd have to still handle edge cases of < 16.  Though in our case I'm not even calling this unless we have 64 bytes due to it being slower otherwise.

However, I believe I just found the source of the bug.  The issue in this circumstance is that algn_diff is 2, not a full 4 bytes.  So when we fold it in, we don't accommodate for the entirety of the initial crc32 checksum.  Hah, that was...tricky.  I'm not sure what the best way to solve it is.  Given that we we call this function, it's at a minimum of 64 bytes, maybe I can account for that by doing unaligned loads out to the next remainder?  Or I could do something a _little_ stranger, like shifting in zeros for the initial value for the XOR.

Reply:
Looks like you found it and it was similar to what I was thinking. What if we do unaligned loads for 16 bytes to load the initial CRC value, before `align_diff` branch?

Reply:
`fold_1` is aligned so we would need an unaligned version possibly...

Reply:
Nice job with the unaligned load. That is simpler than I thought it would be.